### PR TITLE
Assignment4: mmoran314 (Michael Moran) BenFristadSchool (Ben Fristad)

### PR DESCRIPTION
**Pull request by mmoran314**
Added records file with abstract person and other records for student, employee, and book. Added FullName record struct
Added tests for prevously mentioned.

### CODE REVIEW

**Review by hallen10 — COMMENTED** [review]
## Tasks

- Define a full name record (first, last, middle) handling optional and null appropriately. ✔
  - Provide a comment on the full name record on why you selected to define a value or a reference type and ✔
  - Provide a comment on the full name record on why or why not the type is immutable. ✔
- Define an `IEntity` interface: ✔
  - Add an Id property of type `Guid` that is an init-only setter. ✔
  - Add a Name property that is `string`. ✔
- Define an abstract base class that implements `IEntity` - appropriately choosing to implement the interface explicitly or implicitly. ✔
  - Do not implement the `Name` property in this abstract class. ✔
  - Do force any derived classes to provide an implementation for `Name`. ✔
- Define book, student, and employee records - all with entity capabilities. ✔
  - Choose a thoughtful approach for the value of `Name` for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ✔
  - Provide a comment on each interface **member** in each entity explaining why you implemented it implicitly or explicitly. ❌
  - You should consider the relationship between Student and Employee and refactor the common code shared between them. ✔
  - Given the properties on the full name record, you should consider which entities it makes sense to use it. ✔
  - Test that the equality behavior on these entities behaves as expected. ✔

## Fundamentals

- Be sure you enable:
  - Enable Nullability for all projects  ✔
  - Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional) ✔
  - Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
- **Ensure there are no errors or warnings (including code analysis warnings)** ❌ (Getting warnings)
- **All of the above should be unit tested** ✔
- **Choose simplicity over complexity** ✔

Try explaining better on your comments for task such as: "Provide a comment on each interface **member** in each entity explaining why you implemented it implicitly or explicitly"

File: Logger.Tests/RecordTests.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Logger.Tests;
+
+[TestClass]
+public class RecordTests
```
[the remark this answers is not in the document]
Comment (mmoran314) [reply Logger.Tests/RecordTests.cs:11]: They were not returning intended string, all fixed and tested.

File: Logger.Tests/FulllNameTests.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Logger.Tests;
+
+[TestClass]
+public class FullNameTests
```
[the remark this answers is not in the document]
Comment (mmoran314) [reply Logger.Tests/FulllNameTests.cs:11]: Added test

File: Logger/FullName.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Logger;
+// Decided to go with a struct/value type to compare the values incase names are shared.
+public record struct FullName(string First, string Last, string? Middle = null)
```
[the remark this answers is not in the document]
Comment (mmoran314) [reply Logger/FullName.cs:9]: Added

File: Logger/Records.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+{
+    // Implicitly implements IEntity
+    public override string Name { get; } = FullName.ToString();
+}
+
+public record class Book(FullName AuthorName, string BookTitle, string ISBN) : Entity 
+{
+    public string BookTitle { get;} = BookTitle ?? throw new ArgumentNullException(nameof(BookTitle));
+    public string ISBN { get;} = ISBN ?? throw new ArgumentNullException(nameof(ISBN));
+    public FullName AName { get;} = AuthorName;
```
[the remark this answers is not in the document]
Comment (mmoran314) [reply Logger/Records.cs]: Changed

**Pull request closed without merging**